File: LC/NC250/longest_common_prefix.c
// problem:

// Way 1: Vertical Scanning (scan a char across all the strings)
// time - O(N * L); space - O(1)  | N - no of strings, L - length of common prefix
char* longestCommonPrefix(char** strs, int strsSize) {
  if (strs == NULL) return strdup("");

  // pick a char from first string, compare it across all strings
  // if we got a mismatch or string runs out, then return by slicing first str until that index

  for (int p = 0; p < strlen(strs[0]); p++) {
    for (int s = 1; s < strsSize; s++) {

      // if string runs out or there is a mismatch
      if (strs[s][p] == '\0' || strs[0][p] != strs[s][p]) {
        return strndup(strs[0], p);
      }

    }
  }

  // if we reached here then, then entire first string is the longest prefix
  return strs[0];    
}

File: LC/NC250/longest_common_prefix.py
# problem: https://leetcode.com/problems/longest-common-prefix/

# Way 1: Compare & slicing
# time - O(N * L); space - O(1)  | N - no of strings, L - length of common prefix
class Solution:
  def longestCommonPrefix(self, strs: List[str]) -> str:
    # consider first string as common prefix
    prefix = strs[0]

    # start from second element
    for i in range(1, len(strs)):
      # slice last char of prefix, untile it matches the current word
      while not strs[i].startswith(prefix):
        prefix = prefix[:-1]

    return prefix


# Way 2: Vertical Scanning (check each chars of all string)
# time - O(N * M) : N-no of strings, M-len of common prefix, space - O(1)
# same complexity as above method but we avoided slicing of string for each iteration
class Solution:
  def longestCommonPrefix(self, strs: List[str]) -> str:
    if not strs: return ""

    # pick one char from first string and compare it with all strings
    # continue to do this until there is a mismatch
    for i, ch in enumerate(strs[0]):
      for st in strs[1:]:
        if i >= len(st) or ch != st[i]:
          return strs[0][0:i]

    return strs[0]